### PR TITLE
Move PohService and PohRecorder out of banking_stage and into fullnode

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

PohService and PohRecorder are not shared between the TPU and TVU. 

This is part of design for https://solana-labs.github.io/book-edge/leader-validator-transition.html

#### Summary of Changes

* Move the PohRecorder and PohService to the fullnode.
* poh_service always produces ticks

Fixes #

##### Failures

Tracking #2932

tag: @garious

### CODE REVIEW

**Comment by garious** [comment]
Needs a rebase. Those variable names don't exist anymore.

File: src/banking_stage.rs [hunk 1]
```diff
@@ -164,24 +156,26 @@ impl BankingStage {
     }
 
     pub fn process_and_record_transactions(
-        bank: &Bank,
         txs: &[Transaction],
         poh: &PohRecorder,
         working_bank: &WorkingBank,
     ) -> Result<()> {
         let now = Instant::now();
```
Comment (garious) [inline src/banking_stage.rs]:
```rust
let bank = &working_bank.bank;
```

File: benches/banking_stage.rs [hunk 2]
```diff
@@ -39,6 +41,17 @@ fn check_txs(receiver: &Receiver<Vec<Entry>>, ref_tx_count: usize) {
     assert_eq!(total, ref_tx_count);
 }
 
+fn test_recorder(bank: &Arc<Bank>) -> (PohRecorder, Sender<WorkingBank>) {
```
Comment (garious) [inline benches/banking_stage.rs]: Can you prefix functions like this with `create_` as not to confuse them with tests?

**Comment by solana-grimes** [comment]
:broken_heart: Unable to automerge due to CI failure

**Comment by carllin** [comment]
@aeyakovenko, the integration test if failing because when a validator rotates to leader, it's pohrecorder tick height is set to the end of the slot at which the validator was last the leader (some N slots ago). Because the pohrecorder is behind, is going to flush all the ticks from where it starts all the way up to min_tick_height to BroadcastStage, but broadcast is only expecting ticks for the slot for which the node is the leader.

**Comment by aeyakovenko** [comment]
> @aeyakovenko, the integration test is failing because when a validator rotates to leader, it's pohrecorder tick height is set to the end of the slot at which the validator was last the leader (some N slots ago). Because the pohrecorder is behind, is going to flush all the ticks from where it starts all the way up to min_tick_height to BroadcastStage, but broadcast is only expecting ticks for the slot for which the node is the leader.

@carllin then how will the rest of the network receive the missing ticks?

**Comment by carllin** [comment]
@aeyakovenko, those ticks were generated by some other leader. It's the pohrecorder that is behind where the TVU was when the validator transitions to leader.

**Comment by aeyakovenko** [comment]
@garious can you re-approve

File: src/banking_stage.rs [hunk 3]
```diff
@@ -158,16 +146,15 @@ impl BankingStage {
...
             // record and unlock will unlock all the successfull transactions
-            poh.record(hash, processed_transactions, working_bank)?;
+            poh.lock().unwrap().record(hash, processed_transactions)?;
         }
         Ok(())
     }
 
     pub fn process_and_record_transactions(
-        bank: &Bank,
+        bank: &Arc<Bank>,
```
Comment (garious) [inline src/banking_stage.rs]: That looks weird. An artifact of a previous modification?

Comment (aeyakovenko) [reply src/banking_stage.rs]: Yea, I’ll fix tonight

**Comment by aeyakovenko** [comment]
@carllin i added a bunch of tests, and i think fixed a few of the older ones.  can you take a look?

**Comment by aeyakovenko** [comment]
@mvines, checkout my last commit.  I added a step mechanism for ticks.

File: tests/multinode.rs [hunk 4]
```diff
@@ -1976,7 +1994,7 @@ fn test_two_fullnodes_rotate_every_second_tick() {
 
 #[test]
 fn test_one_fullnode_rotate_every_tick_with_transactions() {
-    test_fullnode_rotate(1, 1, false, true);
+    test_fullnode_rotate(2, 2, false, true);
```
Comment (mvines) [inline tests/multinode.rs]: With `2, 2`, the test implementation is now out of sync with the test name `test_one_fullnode_rotate_every_tick_with_transaction`.  Please make them match again

**Comment by garious** [comment]
```
test test_one_fullnode_rotate_every_second_tick ... thread 'solana-broadcaster' panicked at 'assertion failed: self.tick_height + ticks <= self.max_tick_height', src/broadcast_service.rs:114:9
--
  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
  | stack backtrace:
  | 0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
  | at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
  | 1: std::sys_common::backtrace::_print
  | at src/libstd/sys_common/backtrace.rs:71
  | 2: std::panicking::default_hook::{{closure}}
  | at src/libstd/sys_common/backtrace.rs:59
  | at src/libstd/panicking.rs:211
  | 3: std::panicking::default_hook
  | at src/libstd/panicking.rs:227
  | 4: std::panicking::rust_panic_with_hook
  | at src/libstd/panicking.rs:491
  | 5: std::panicking::begin_panic
  | at /rustc/9fda7c2237db910e41d6a712e9a2139b352e558b/src/libstd/panicking.rs:425
  | 6: solana::broadcast_service::Broadcast::run
  | at src/broadcast_service.rs:114
  | 7: solana::broadcast_service::BroadcastService::run
  | at src/broadcast_service.rs:243
  | 8: solana::broadcast_service::BroadcastService::new::{{closure}}
  | at src/broadcast_service.rs:298
```

**Comment by solana-grimes** [comment]
:broken_heart: Unable to automerge due to CI failure

**Comment by solana-grimes** [comment]
:broken_heart: Unable to automerge due to CI failure

**Pull request merged**